plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskContext.java

```diff
@@ -62,6 +60,22 @@
 @SuppressWarnings({ "restriction" })
 public class ResourceTaskContext {
 
+	/*
+	 * Review feedback:
+	 *
+	 * It *appears* as if we are duplicating quite some logic in this class. The builder front-end is responsible for
+	 * updating resources after changes happened. We are implementing a different incremental build semantics in
+	 * refreshContext with fewer optimizations and therefore a performance drag.
+	 *
+	 * The logic, how a given resource set is updated after an incoming change is recorded, should certainly be
+	 * implemented exactly once.
+	 *
+	 * For xtext.ui, we made the mistake of duplicating this between the Eclipse builder and the reconciler, and this
+	 * was a constant PITA.
+	 *
+	 * #prepareRefresh is sort-of a brute force approach that has potential for optimizations.
```

Comment:
Fair point indeed. The Eclipse builder can simply "make a mess" with its current resource set in any case, since it will be discarded after the build was done. The drawback of this is the high setup cost for each build run. In Eclipse, we rely on the fact that the AutoBuildJob will only be triggered with certain delay which mitigates this cost-issue somehow.

Measurements in other language context showed though, that we have extremely short built times in LSP due to the reused resource set. That is something we should really try to keep as a non-functional property.